design/one-pager-consuming-k8s-clusters.md

```diff
@@ -0,0 +1,270 @@
+   that references the previously created Kubernetes `Provider`.
+1. Application owner provisions a `KubernetesApplication` with labels that
+   select that `WorkloadTarget` for provisioning.
+
+This example shows the power of this model, but also one of the notable
+shortcomings: if a user can publish clusters to a namespace with a
+`WorkloadTarget`, then they can consume any cluster. For the application owners
+that are doing self-service provisioning in this scenario, they will likely need
+to be able to create `WorkloadTarget` resources for their dynamically
+provisioned Kubernetes cluster managed resources. However, if they can do so,
+they could just as easily create a `WorkloadTarget` for the Kubernetes
+`Provider` that is only to be used for sensitive workloads. While it would be
+nice to restrict the clusters that could be published by users with the ability
+to create `WorkloadTarget` resources, there is not an immediately clear and
+elegant way to do so. It could be argued that only trusted users should have
+this permission level anyway, so it may be less of an issue than expressed here.
+Nevertheless, a potential solution is outlined [below](#future-considerations).
+
+## Technical Implementation
+
+The implementation of this proposal is relatively lightweight and could
+potentially be implemented without churn on any existing exposed APIs. The first
+step would be implementing the `WorkloadTarget` type in [core Crossplane]. It
+would not require any additional controllers.
+
+The next step would be to modify the `KubernetesApplication` [scheduling
+controller] to list `WorkloadTarget` resources in the namespace that match
+labels, instead of the current implementation that lists `KubernetesCluster`
+claims. Currently, this controller sets a reference to a `KubernetesCluster`
+claim, which is then later propagated to the child
+`KubernetesApplicationResource` objects, which use it to get a `Secret` with
+[remote cluster connection information].
+
+Because a `WorkloadTarget` may reference either a Kubernetes cluster managed
+resource or a Kubernetes `Provider` resource, both of which ultimately lead to a
+connection `Secret`, it would likely be cleanest to go ahead and get that
+`Secret` and set a reference to it on the `KubernetesApplication`, then
+propagate to the `KubernetesApplicationResource` objects. The alternative option
+would be to set a reference to either the managed resource or the `Provider`,
+then follow the references to the corresponding `Secret` in the remote cluster
+connection step in the `KubernetesApplicationResource` object controller. While
+setting the reference to the `Secret` is potentially a cross-namespace
+reference, it is one that only exists in the status of the resource, and a valid
+`WorkloadTarget` would have to exist in the `KubernetesApplication` namespace
+for that cross-namespace reference to be set by the controller.
+
+## Future Considerations
```

Comment:
I think having a namespace-pointing `WorkloadTarget` could be a future consideration. As infra-owner, when I give access to two different teams, I trust them to not mess with each other's resources. But if I can limit their access to 1 namespace in the target cluster, it'd be much safer for all parties.

Reply:
That is an interesting idea for sure! I think it might be cool to even create a `Role` / `RoleBinding` / `ServiceAccount` to establish RBAC permissions for deploying as well. I definitely think that would be a future consideration because I am not certain how much it would actually be used in practice in a true "cluster as cattle" scenario where the cluster is the unit of isolation.